### ISSUE TITLE
Implicit return and arrow functions

### ISSUE THREAD

**Issue by gabro** [buildo/eslint-config#18]
I find this confusing:

``` js
[1, 2, 3].map((foo) =>
  bar.something(blabla).foo()
);
```

That's an implicit return, but I personally can't parse it easily because it's on a separate line.

I propose to always stick to either

``` js
[1, 2, 3].map((foo) => bar.something(blabla).foo());
```

or 

``` js
[1, 2, 3].map((foo) => {
  return bar.something(blabla).foo();
});
```

This improves readability and predictability, imho. @buildo/frontend, thoughts?

**Comment by francescogior** [comment]
I like to use implicit return for functions that returns 'templates', even if split in lines

``` js
specialTemplate = ({ arg1, arg2, ...args }) => (
  <div>
    <EhiComeVa />
    <BeneTe />
    <UnPoDiAcchiacchi />
    <MiSpiace />
  </div>
)
```

**Comment by gabro** [comment]
good point, I think that's acceptable.
It's not a matter of single/multiple lines, rather whether the expression is "linked" to the line where `=>` is.

New formulation of the "rule":

> implicit returned expression of arrow functions must begin on the same line of the arrow.

Another similar example

``` js
({ foo, bar }) => `
  Hey ${foo}
  ci vediamo al ${bar}?
`
```

**Comment by ascariandrea** [comment]
I don't think you resolve readability on last example by make the return explicit:

``` js
({ foo, bar }) => {
 return `
  Hey ${foo}
  ci vediamo al ${bar}?
`;
}
```

What's the fat arrow implicit return purpose? Avoid the overhead of writing return statement for functions with single declaration in body?

I think we should encourage that as long as is readable.

``` js
arrayOfArrays.map(x =>
 x.filter(y => someCondition(y))
  .map(y => y * 2)
)
```

**Comment by gabro** [comment]
Sorry, I realize I wasn't clear.

The last example I posted is a **good** one, in my view. The implicit return is made clear by the fact that the expression begins right after the `=>`.

The last you posted is not, in my opinion. I find it hard to parse, because of the new line after the `=>`

**Comment by gabro** [comment]
To sum it up

## Good cases

``` js
[1, 2, 3].map(i => i);
```

``` js
[1, 2, 3].map(i => {
  return i;
});
```

```
[1, 2, 3].map(i => `
  ${i}
`);
```

``` js
[1, 2, 3].map(i => (
  <span>
    ${i}
  </span>
);
```

## Bad cases

``` js
[1, 2, 3].map(i =>
  i
);
```

``` js
[1, 2, 3].map(i =>
  (
    <span>
      {i}
    </span>
  )
);
```

**Comment by giogonzo** [comment]
> implicit returned expression of arrow functions must begin on the same line of the arrow

I'd say this sums it up perfectly, :+1: 

But I don't think there's any rule for that (slightly related: http://eslint.org/docs/rules/arrow-body-style)

**Comment by gabro** [comment]
:angel: https://github.com/buildo/eslint-plugin-fat-arrow-same-line

**Comment by gabro** [comment]
Since there's no agreement on this, I'll close it for now
